### PR TITLE
Decouple backend into RISCV and ARM

### PR DESCRIPTION
**Pull request by eecheng87**
This patch is prepared for importing RISC-V backend in future.
Config variable in Makefile to switch target machine,
also need to config included backend file in main.c

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
How about adding new target "make config"? It can configure the specific targets. The sample usage would be "make config" (Arm target), "make config ARCH=arm" (explicit Arm target) and "make config ARCH=riscv" (explicit RISC-V target)

**Comment by eecheng87** [comment]
I've added new way to switch between modes.

Quick demo:
```shell
Demo:

$ make config ARCH=arm
Target machine code switch to arm

$ make
  CC+LD	out/inliner
  GEN	out/libc.inc
  CC	out/src/main.o
  LD	out/shecc
  SHECC	out/shecc-stage1.elf
  SHECC	out/shecc-stage2.elf
```

If it looks good for you, please tell me. I'll also change README about configuration (maybe in this or next PR)

**Comment by jserv** [comment]
> If it looks good for you, please tell me. I'll also change README about configuration (maybe in this or next PR)

It looks great. Thanks. Let's work within this pull request.

File: tests/driver.sh [hunk 1]
```diff
@@ -10,11 +10,11 @@ function try() {
 
     local tmp_in="$(mktemp --suffix .c)"
     local tmp_exe="$(mktemp)"
-
+    local target_exe=$(<"$PWD/out/target")
     echo "$input" > "$tmp_in"
     "$SHECC" -o "$tmp_exe" "$tmp_in"
     chmod +x $tmp_exe
-    $ARM_EXEC "$tmp_exe"
+    "$target_exe" "$tmp_exe"
```
Comment (jserv) [inline tests/driver.sh:17]: should be `$target_exec`

File: README.md [hunk 2]
```diff
@@ -59,6 +59,15 @@ the second stage bootstrapping would fail due to `qemu-arm` absence.
 
 ## Build and Verify
 
+Configure which backend you want, `shecc` supports `ARMv7-A` and `RISC-V32I` backend:
+```
+$ make config ARCH=arm
+Target machine code switch to arm
```
Comment (jserv) [inline README.md:65]: Should be `# Target machine code switch to Arm`

File: README.md [hunk 3]
```diff
@@ -59,6 +59,15 @@ the second stage bootstrapping would fail due to `qemu-arm` absence.
...
+
+$ make config ARCH=riscv
+Target machine code switch to riscv
```
Comment (jserv) [inline README.md:68]: Ditto.

File: tests/driver.sh [hunk 4]
```diff
@@ -10,11 +10,11 @@ function try() {
 
     local tmp_in="$(mktemp --suffix .c)"
     local tmp_exe="$(mktemp)"
-
+    local target_exe=$(<"$PWD/out/target")
     echo "$input" > "$tmp_in"
     "$SHECC" -o "$tmp_exe" "$tmp_in"
     chmod +x $tmp_exe
-    $ARM_EXEC "$tmp_exe"
+    $target_exe "$tmp_exe"
```
Comment (jserv) [inline tests/driver.sh:17]: Should be `$target_exec`

File: src/main.c [hunk 5]
```diff
@@ -16,8 +16,9 @@
 /* C language front-end */
 #include "cfront.c"
 
-/* Machine code generation. ARMv7-A is the only target */
-#include "codegen.c"
+/* Machine code generation. support ARMv7-A and RISC-V32I */
+#include "arm-codegen.c"
+/* include "riscv-codegen.c" */
```
Comment (jserv) [inline src/main.c:21]: Remove `riscv-codegen.c` here.

**Pull request merged**